revamp prior api

francescoalemanno opened the pull request:
Ty Jerry for all the PRs, and all the work you're putting into this.
@Moelf 
This PR is supposed to do this:
1) Implement the original prior in Scargle's paper + Scargle's correction. i.e. the prior is supposed to be evaluated using the total number of possible splits `N` :
`ncp_prior = 4 - log( fp_rate / ( 0.0136 * length(t) .^ (0.478 ) ) );`
2) <strike>Implement a non-informative prior on the single within-bin counts this is `Jeffrey(0)`, it looks very reasonable on some test data</strike>
implement a easy to use prior like Pearson(p) and Geometric(gamma)
3) Provide an API which is flexible enough to allow users to plug whatever they want.

does this look reasonable?

TODO's:
1) maybe `Scargle(0.05)` is no longer a good default.
2) maybe another kind of prior besides Jeffrey on bin counts must be implemented and offered as default.

Moelf commented:
>maybe Scargle(0.05) is no longer a good default.

I think the geometric prior is pretty good, regardless of what the heck data is, `N_b << N` is a good assumption anyway

Moelf commented on src/BayesHistogram.jl:
```diff
@@ -33,12 +33,33 @@ end
-        (cnt_total * sqrt(cnt_total) / 2) /
-        (sqrt(cnt_single) * (C0 + cnt_total * (1 / 4 + C1 * sqrt(cnt_total) + cnt_total))),
-    )
+    ### MATHEMATICA CODE
```
big brain moment

francescoalemanno replied:
🤣

francescoalemanno commented:
I added you as a collaborator btw :)

Moelf commented:
oh yeah we can totally just do this:
![image](https://user-images.githubusercontent.com/5306213/197844437-ca6e2bbd-5e10-4da2-a7a8-7912e94654b5.png)

is this what we need?
```diff
diff --git a/src/BayesHistogram.jl b/src/BayesHistogram.jl
index 1152ef8..d85314e 100644
--- a/src/BayesHistogram.jl
+++ b/src/BayesHistogram.jl
@@ -41,6 +41,16 @@ function (w::Scargle)(n, n_max)
     log(C0*w.p0*n^C1) - 4.0
 end

+struct Geometric{T<:Real, W<:AbstractFloat}
+    gamma::T
+end
+
+function (w::Geometric)(n, n_max)
+    gamma = w.gamma
+    N_p1 = n_max+1
+    (1-gamma)/(1-gamma^N_p1) * gamma^(n)
+end
+
```

Moelf commented:
oh wait:
```> But Eq. (3) is very convenient
to implement, since with the fitness equal to the log of the posterior,
one only needs to subtract the constant −log γ (called ncp_prior
in the MatLab code

francescoalemanno commented:
True, I also at some point realised that there is no need for normalisation constants :)
btw I removed that "Jeffrey" monstrosity and added a simple Pearson Chi2 prior and the Geometric prior

Moelf commented:
![image](https://user-images.githubusercontent.com/5306213/197877613-4bfc0c24-6a53-40d4-9b31-165aa6fc868f.png)

you didn't update plots it seems

francescoalemanno commented:
> ![image](https://user-images.githubusercontent.com/5306213/197877613-4bfc0c24-6a53-40d4-9b31-165aa6fc868f.png)
> 
> you didn't update plots it seems

now I fixed it, at some point the geometric distribution was defined without a minus so the gamma had to be small, now it has to be large 👍🏼

Moelf commented:
hmmm, is there a way to normalize this so gamma can just be from 0 to 1? or no

Moelf commented:
wait yeah that looks great and we can use the inner constructor to limit gamma to be 0-1

Moelf commented:
ok I'm really happy with what we have now feel free to merge
